plugins/services/src/js/reducers/serviceForm/JSONReducers/resourceLimits.ts

```diff
@@ -0,0 +1,46 @@
+function resourceLimitReducer(resourceField: string, parseFn = parseInt) {
+  return function(
+    this: { cache?: number | string },
+    state: string | number | null,
+    args: { path: string; value: any } = { path: "", value: null }
+  ) {
+    const { path, value } = args;
+    const [limit, resource] =
+      path.slice(-3)[0] === "limits" ? path.slice(-3) : path.slice(-2);
+    const numberValue = parseFn(value);
+    if (limit !== "limits" || resourceField !== resource) {
+      return state;
+    }
+    if (isNaN(numberValue)) {
+      return value === false
+        ? this.cache || null
+        : (value === "unlimited" || value === true) && "unlimited";
+    }
+    this.cache = numberValue;
+    return this.cache;
+  };
+}
+
+export function JSONReducer(
+  this: {
+    mem: { cache?: number | string };
+    cpu: { cache?: number | string };
+  },
+  state = { cpus: null, mem: null },
+  { path, value }
+) {
+  this.mem = this.mem == null ? {} : this.mem;
```

Comment:
`this.mem == null` => `!this.mem`, right?

Reply:
We are basically initialising the cache variables, if they are not defined yet.
```JS
if(this.mem == null) {
  this.mem = {}
}

// Would it be more obvious to do it like this

this = {...{cpus:{}, mem:{}}, ...this};